packages/react-intersection-observer/src/hooks.ts

```diff
@@ -0,0 +1,147 @@
+  root,
+  rootMargin,
+  threshold,
+  unsupportedBehavior = UnsupportedBehavior.TreatAsIntersecting,
+}: Options = {}): [IntersectionObserverEntry, Ref<HTMLElement | null>] {
+  const node = useRef<HTMLElement | null>(null);
+  const lastNode = useRef<HTMLElement | null>(null);
+  const observer = useRef<IntersectionObserver | null>(null);
+  const lastObserver = useRef<IntersectionObserver | null>(null);
+
+  const [intersectionEntry, setIntersectingEntry] = useState<
+    IntersectionObserverEntry
+  >(() => ({
+    boundingClientRect: emptyBoundingClientRect,
+    intersectionRatio: 0,
+    intersectionRect: emptyBoundingClientRect,
+    isIntersecting: false,
+    rootBounds: emptyBoundingClientRect,
+    target: null as any,
+    time: Date.now(),
+  }));
+
+  useEffect(
+    () => {
+      if (!isSupported()) {
+        return;
+      }
+
+      const resolvedRoot =
+        typeof root === 'string' ? document.querySelector(root) : root;
+
+      const intersectionObserver = new IntersectionObserver(
+        ([entry]) =>
+          setIntersectingEntry({
+            ...entry,
+            // Normalizes for inconsistent browser support
+            isIntersecting: entry.intersectionRatio > 0,
+          }),
+        {
+          root: resolvedRoot,
+          rootMargin,
+          threshold,
+        },
+      );
+
+      observer.current = intersectionObserver;
+
+      // eslint-disable-next-line consistent-return
+      return () => {
+        intersectionObserver.disconnect();
+      };
+    },
+    [root, rootMargin, Array.isArray(threshold) ? threshold.join() : threshold],
+  );
+
+  useEffect(() => {
+    if (
+      lastNode.current === node.current &&
+      lastObserver.current === observer.current
+    ) {
+      return;
+    }
+
+    lastNode.current = node.current;
+
+    if (node.current == null) {
+      return;
+    }
+
+    if (
+      !isSupported() &&
+      unsupportedBehavior === UnsupportedBehavior.TreatAsIntersecting
+    ) {
+      const boundingClientRect = node.current.getBoundingClientRect();
+
+      setIntersectingEntry({
+        boundingClientRect,
+        intersectionRatio: 1,
+        intersectionRect: boundingClientRect,
+        isIntersecting: true,
+        rootBounds: boundingClientRect,
+        target: node.current,
+        time: Date.now(),
+      });
+
+      return;
+    }
+
+    if (observer.current != null) {
+      lastObserver.current = observer.current;
+      observer.current.observe(node.current);
+    }
+
+    // eslint-disable-next-line consistent-return
+    return () => {
+      if (
+        lastNode.current == null ||
+        lastObserver.current == null ||
+        (lastNode.current === node.current &&
+          lastObserver.current === observer.current)
+      ) {
+        return;
+      }
+
+      lastObserver.current.unobserve(lastNode.current);
+    };
+  });
+
+  return [intersectionEntry, node];
+}
+
+export function useValueTracking<T>(
```

Comment:
this makes sense as a shared hook https://github.com/Shopify/quilt/pull/609/files#diff-99a7b9f41d8c46ef5eb67c2b0da03b2aR3

Reply:
agreed

Reply:
Yes, agreed, we just don't have a place for them yet, and I didn't want to throw that in here too.